docs/specs/pipelines-io.md

```diff
@@ -0,0 +1,196 @@
+## Efficient IO
+
+As part of the high performance work, we need new abstractions for doing efficient reading and writing to and from the network. The pattern we have settled on
+so far in pipelines is a model where buffers are pushed by a producer (`IPipelineWriter`) to a consumer (`IPipelineReader`).
+
+```C#
+public interface IPipelineReader
+{
+    ReadableBufferAwaitable ReadAsync();
+    void Advance(ReadCursor consumed, ReadCursor examined);
+    void CancelPendingRead();
+    void Complete(Exception exception = null);
+}
+```
+
+```C#
+public interface IPipelineWriter
+{
+    Task Writing { get; }
+    WritableBuffer Alloc(int minimumSize = 0);
+    void Complete(Exception exception = null);
+}
+```
+
+The producer gets a handle on a `WritableBuffer` and the consumer gets a handle on a `ReadableBuffer` to do writes and reads respectively. 
+
+Memory in pipelines is managed as a linked list of `OwnedMemory<T>`:
+
+```C#
+public class BufferSegment
```

Comment:
Consider using `sealed` here.  Given this is high performance code and JIT in some cases can take advantage of sealed it should be considered.